### PR TITLE
Refactor Parameter Generation

### LINKED ISSUE #149: Generated Parameters should be partial replacement.
Keyword based parameter values do not support partial replacement. (those generated by `$[taskcat_*] `keywords). 

A better value-add here is to support partial replacement. The proposed changes are backwards compatible, as parameter values that only contain the keywords will be fully replaced. However, parameter values that contain a keyword, plus additional text, will be partially replaced. 

Ex: 
```
 {
        "ParameterKey": "AWSPublicFQDN",
        "ParameterValue": "foobar-$[taskcat_random-string].example.com"
 }
```

Would evaluate to:

```
 {
        "ParameterKey": "AWSPublicFQDN",
        "ParameterValue": "foobar-slkdfjsdfslkdflk.example.com"
 }
```

Non-inclusive examples of where this should be changed.

https://github.com/aws-quickstart/taskcat/blob/c86818c16eabd25852afcef1112574c97fe179c4/taskcat/stacker.py#L895-L897
https://github.com/aws-quickstart/taskcat/blob/c86818c16eabd25852afcef1112574c97fe179c4/taskcat/stacker.py#L903-L905
(...)

### PR DESCRIPTION
**Pull request by agamzn**
A refactor of parameter generation. Unit Tests have been added that verify, among other things: 
- All Regular expressions match the placeholder they're intended.
- No regular expressions are added without a unit test.
- Wrapper functions are tested.
- Parameter generation function is tested.

Fixes #149 ;

Potentially supersedes: #224

### CODE REVIEW

File: taskcat/stacker.py [hunk 1]
```diff
@@ -794,237 +670,8 @@ def generate_input_param_values(self, s_parms, region):
...
-
-                if getval_re.search(param_value):
-                    requested_key = self.regxfind(getval_re, param_value)
-                    print("{}Getting previously assigned value for {}".format(PrintMsg.DEBUG, requested_key))
-                    param_value = self.get_parameter(requested_key)
-                    print("{}Loading {} as value for {} ".format(PrintMsg.DEBUG, param_value, requested_key))
-                    _parameters['ParameterValue'] = param_value
-
-        return s_parms
+        x = ParamGen(param_list=s_parms, bucket_name=self.get_s3bucket(), boto_client=self._boto_client, region=region, verbose=True)
```
Comment (latrokles) [inline taskcat/stacker.py]: the tmp var x is not being used anywhere else. we could:
`return ParamGen(...).param_list`

File: taskcat/template_params.py [hunk 2]
```diff
@@ -0,0 +1,361 @@
+import re
+import random
+import uuid
+import logging
+from taskcat.colored_console import PrintMsg
+from taskcat.exceptions import TaskCatException
+
+logger = logging.getLogger('taskcat')
+class ParamGen(object):
```
Comment (latrokles) [inline taskcat/template_params.py]: we no longer have to subclass object in python.
```suggestion
 class ParamGen:
```

File: taskcat/template_params.py [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+        s3_client = self._boto_client.get('s3', region=self.region, s3v4=True)
+        try:
+            dict_object = s3_client.get_object(Bucket=bucket, Key=object_key)
+        except TaskCatException:
+            raise
+        except Exception:
+            print("{} Attempted to fetch Bucket: {}, Key: {}".format(PrintMsg.ERROR, bucket, object_key))
+            raise
+        content = dict_object['Body'].read().decode('utf-8').strip()
+        return content
```
Comment (latrokles) [inline taskcat/template_params.py]: Looking at `ClientFactory.get` it too can raise TaskCatException, so it'd be good to place inside the try.

Also, given that the `return` statements are only executed if there are no exceptions, these should also be inside the `try` block. This prevents any dangling logic and clarifies the execution flow:
```
def my_function(params):
  try:
    result = compute_some_stuff_or_get_stuff
    the_thing_that_matters = extract_thing_from_result
    return the_thing_that_matters
  except SomeException:
    raise
```
Reading this I can see what the happy path is right away and then move on to the not-happy path by seeing how errors are handled.

On a related note:
1. if we are propagating the exceptions back up to the callee... does it matter to handle them?
2. if we are handling them for logging purposes, shouldn't we log the TaskCatException as well?
3. if we are going to log the errors, should we also log the message from the exception (maybe even as a debug or trace?)

File: taskcat/template_params.py [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+
+    def genpassword(self, pass_length, pass_type=None):
```
Comment (latrokles) [inline taskcat/template_params.py]: There's enough logic related to password gen that I could see this becoming its own little class. Not necessarily as part of this change, but something to think about for the future.

File: tests/unittest/test_template_params.py [hunk 5]
```diff
@@ -0,0 +1,314 @@
+import unittest
```
Comment (latrokles) [inline tests/unittest/test_template_params.py]: same thing here about `(object):` not being needed.

File: taskcat/template_params.py (hunk 4 above)
Comment (agamzn) [reply taskcat/template_params.py]: Out of scope for now, but certainly OK with revisiting later.

**Comment by agamzn** [comment]
@latrokles  - new pushes to reflect your comments.

File: taskcat/template_params.py (hunk 3 above)
Comment (latrokles) [reply taskcat/template_params.py]: that's fair. We can get this merged and then look into a future change that tackles this bit in isolation.

would like to note that your concern is in itself a good argument to look into this method further.

File: taskcat/template_params.py (hunk 4 above)
Comment (latrokles) [reply taskcat/template_params.py]: 👍

**Comment by vsnyc** [comment]
looks great!

File: taskcat/stacker.py [hunk 6]
```diff
@@ -1115,9 +761,10 @@ def stackcreate(self, taskcat_cfg, test_list, sprefix):
                 except TaskCatException:
                     raise
                 except Exception as e:
-                    if self.verbose:
-                        print(PrintMsg.ERROR + str(e))
-                    raise TaskCatException("Cannot launch %s" % self.get_template_file())
+                    raise
+#                    if self.verbose:
```
Comment (latrokles) [inline taskcat/stacker.py]: we should move away from committing commented out code. we can always pull code back from version history, it's part of the point.

**Pull request merged**